docs/cub/developer_overview.rst

```diff
@@ -386,282 +386,264 @@ Device-scope
+    __launch_bounds__(ChainedPolicy::ActivePolicy::AlgorithmPolicy::BLOCK_THREADS) CUB_DETAIL_KERNEL_ATTRIBUTES
+    void kernel(...) { // (4)
+      using policy = ChainedPolicy::ActivePolicy; // selects policy of active device compilation pass (5)
+      using agent = AgentAlgorithm<policy>; // instantiates (6)
+      agent a{...};
+      a.Process(); // calls (7)
     }
 
-    template <int PTX_VERSION, typename PolicyT, typename PrevPolicyT>
-    struct ChainedPolicy {
-      using ActivePolicy = conditional_t<(CUB_PTX_ARCH < PTX_VERSION), // (5)
-                                        PrevPolicyT::ActivePolicy,
-                                        PolicyT>;
+    template <typename Policy>
+    struct AlgorithmAgent {  // (6)
+      void Process() { ... } // (7)
     };
 
-    template <class Policy>
-    struct AlgorithmAgent { // (6)
-      void Process();
-    };
-
-The code above represents control flow.
 Let's look at each of the building blocks closer.
 
-The dispatch entry point is typically represented by a static member function that constructs an object of ``DispatchReduce`` and passes it to ``ChainedPolicy`` ``Invoke`` method:
+The dispatch entry point is typically represented by a static member function called ``DispatchAlgorithm::Dispatch``
+that constructs an object of type ``DispatchAlgorithm``, filling it with all arguments to run the algorithm,
+and passes it to the ``ChainedPolicy::Invoke`` function:
 
 .. code-block:: c++
 
-    template <typename InputIteratorT,
-              // ...
-              typename SelectedPolicy>
-    struct DispatchReduce : SelectedPolicy
-    {
-
-      //
-      CUB_RUNTIME_FUNCTION __forceinline__ static
-      cudaError_t Dispatch(
-          void *d_temp_storage, size_t &temp_storage_bytes,
-          InputIteratorT d_in,
-          /* ... */)
-      {
-        typedef typename DispatchSegmentedReduce::MaxPolicy MaxPolicyT;
-
-        if (num_segments <= 0) return cudaSuccess;
-
-        cudaError error = cudaSuccess;
-        do {
-          // Get PTX version
-          int ptx_version = 0;
-          error = CubDebug(PtxVersion(ptx_version));
-          if (cudaSuccess != error)
-          {
-            break;
-          }
-
-          // Create dispatch functor
-          DispatchSegmentedReduce dispatch(
-              d_temp_storage, temp_storage_bytes, d_in, /* ... */);
-
-          // Dispatch to chained policy
-          MaxPolicyT::Invoke(ptx_version, dispatch);
-        } while (0);
-
-        return error;
+    template <..., // algorithm specific compile-time parameters
+              typename SelectedPolicy> // also called: PolicyHub
+    struct DispatchAlgorithm : SelectedPolicy { // TODO(bgruber): I see no need for inheritance, can we remove it?
+      CUB_RUNTIME_FUNCTION _CCCL_FORCEINLINE static
+      cudaError_t Dispatch(void *d_temp_storage, size_t &temp_storage_bytes, ..., cudaStream stream) {
+        if (/* nothing no items to process */) return cudaSuccess;
```

Comment:
important: the contract is that we set the temporary storage bytes. As written, user might provide uninitialized temporary storage bytes containing a large value, first phase returns success error code, user allocates temporary storage of undefined size and potentially gets OOM. I must've missed that while writing this overview. Let's fix this while we are here to something along the lines of:
```suggestion
        if (/* no items to process */) {
          if (d_temp_storage == nullptr) {
            temp_storage_bytes = 1;
          }
          return cudaSuccess;
        }
```

Reply:
I took it from the previous version of the docs, but I was thinking the same: we should set the `temp_storage_bytes`.